src/Compilers/CSharp/Test/Semantic/Semantics/UsingStatementTests.cs

```diff
@@ -1452,6 +1453,280 @@ static void Main()
+    {
+        await using (C1 c = new C1())
+        {
+        }
+    }
+}";
+            var compilation = CreateCompilationWithTasksExtensions(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(compilation, expectedOutput: "Dispose async");
+        }
+
+        [Fact]
+        public void UsingPatternAsyncWithTaskLikeReturnTest()
+        {
+            var source = @"
+using System;
+using System.Runtime.CompilerServices;
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+class C1
+{
+    public C1() { }
+    public System.Threading.Tasks.Task DisposeAsync() { return System.Threading.Tasks.Task.CompletedTask; }
+}
+
+class C2
+{
+    public C2() { }
+    public MyTask DisposeAsync() { return new MyTask(); }
+}
+
+[AsyncMethodBuilder(typeof(MyTaskMethodBuilder))]
+struct MyTask
+{
+    internal Awaiter GetAwaiter() => new Awaiter();
+    internal class Awaiter : INotifyCompletion
+    {
+        public void OnCompleted(Action a) { }
+        internal bool IsCompleted => true;
+        internal void GetResult() { }
+    }
+}
+
+struct MyTaskMethodBuilder
+{
+    private MyTask _task;
+    public static MyTaskMethodBuilder Create() => new MyTaskMethodBuilder(new MyTask());
+    internal MyTaskMethodBuilder(MyTask task)
+    {
+        _task = task;
+    }
+    public void SetStateMachine(IAsyncStateMachine stateMachine) { }
+    public void Start<TStateMachine>(ref TStateMachine stateMachine) where TStateMachine : IAsyncStateMachine
+    {
+        stateMachine.MoveNext();
+    }
+    public void SetException(Exception e) { }
+    public void SetResult() { }
+    public void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : INotifyCompletion where TStateMachine : IAsyncStateMachine { }
+    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : ICriticalNotifyCompletion where TStateMachine : IAsyncStateMachine { }
+    public MyTask Task => _task;
+}
+
+
+
+class C3
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        await using (C1 c = new C1())
+        {
+        }
+
+        await using (C2 c = new C2())
+        {
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void UsingPatternAsyncNoAwaitTest()
+        {
+            var source = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+class C1
+{
+    public C1() { }
+    public System.Threading.Tasks.ValueTask DisposeAsync() { return new System.Threading.Tasks.ValueTask(System.Threading.Tasks.Task.CompletedTask); }
+}
+
+class C2
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        using (C1 c = new C1())
```

Comment:
>using (C1 c = new C1()) [](start = 8, length = 23)

It looks like this PR is based on a slightly aged bits from preview2. Preview2 now has a special message for such cases: `ERR_NoConvToIAsyncDispWrongAsync`. 
It's basically the same message, but adds "Did you mean to use `await using` instead of `using`?"
FYI for when you integrate